crates/server/src/session.rs

```diff
@@ -0,0 +1,135 @@
+//! Data model, state management, and configuration resolution.
+
+use lsp_types::Url;
+use lsp_types::WorkspaceFolder;
+
+use crate::document::{DocumentKey, DocumentVersion, PositionEncoding, TextDocument};
+
+pub(crate) use self::capabilities::ResolvedClientCapabilities;
+pub use self::index::DocumentQuery;
+
+mod capabilities;
+mod index;
+mod workspaces;
+
+/// The global state for the LSP
+pub(crate) struct Session {
+    /// Used to retrieve information about open documents and settings.
+    index: index::Index,
+    /// The global position encoding, negotiated during LSP initialization.
+    position_encoding: PositionEncoding,
+    /// Tracks what LSP features the client supports and doesn't support.
+    resolved_client_capabilities: ResolvedClientCapabilities,
+}
+
+/// An immutable snapshot of `Session` that references
+/// a specific document.
+pub(crate) struct DocumentSnapshot {
+    document_ref: index::DocumentQuery,
+    position_encoding: PositionEncoding,
+}
```

Comment:
The `Session` is like our `LspState` and the `DocumentSnapshot` is like our `WorldState`

We can expand on these as needed going forward, but should be careful about what we put in `DocumentSnapshot`